### PR TITLE
Handle and trace errors in secret stores

### LINKED ISSUE #1070: SecretStore should not swallow errors while retrieving secrets
Existing trait signature does not allow underline Secret Store implementation to report error while getting a secret. 

https://github.com/spiceai/spiceai/blob/ccf8615eca63f483ae82c6194da28c255a548eb6/crates/secrets/src/lib.rs#L42
```
#[async_trait]
pub trait SecretStore {
    async fn get_secret(&self, secret_name: &str) -> Option<Secret>;
}
```

As a results those errors are ignored/shallowed that makes troubleshooting secrets related errors very complicated. That doesn't allow the caller to also distinguish between whether a secret isn't there, or there was some error getting the secret. We should update Api to allow passing error which should be processed in an centralized way (log ged as WARN for example)

```
async fn get_secret(&self, secret_name: &str) -> Option<Secret> {
        if let Ok(secret) = self.kubernetes_client.get_secret(secret_name).await {
            return Some(Secret::new(secret.clone()));
        }

        None
    }
```

### PR DESCRIPTION
**Pull request by ewgenius**
closes #1070 

Adds error handling and tracing for secret stores. Handles all suitable errors, swallowing them only when the secret is explicitly known to be none. Properly passes errors down.

Examples:

CleanShot 2024-04-22 at 23 59 14@2x

CleanShot 2024-04-22 at 23 59 34@2x

CleanShot 2024-04-23 at 00 00 00@2x

CleanShot 2024-04-23 at 00 02 02@2x

CleanShot 2024-04-23 at 00 01 02@2x

### CODE REVIEW

File: crates/secrets/src/aws_secrets_manager.rs [hunk 1]
```diff
@@ -110,29 +110,29 @@ impl SecretStore for AwsSecretsManager {
...
+                return Ok(None);
             }
             Err(err) => {
                 tracing::warn!(
                     "Failed to get secret {} from AWS Secrets Manager: {}",
                     secret_name,
                     err
                 );
-                return None;
+                return Ok(None);
```
Comment (phillipleblanc) [inline crates/secrets/src/aws_secrets_manager.rs]: Shouldn't this return an `Err`?

File: crates/secrets/src/keyring.rs [hunk 2]
```diff
@@ -42,23 +42,54 @@ impl KeyringSecretStore {
...
+
+        let entry = match Entry::new(entry_key.as_str(), "spiced") {
+            Ok(entry) => entry,
+            Err(err) => {
+                tracing::warn!(
+                    "Failed to get secret entry {} from keyring store, {}",
+                    entry_key,
+                    err
+                );
+                return Ok(None);
```
Comment (phillipleblanc) [inline crates/secrets/src/keyring.rs:67]: Shouldn't this return an `Err`?

Comment (ewgenius) [reply crates/secrets/src/keyring.rs:67]: fixed

File: crates/secrets/src/aws_secrets_manager.rs (hunk 1 above)
Comment (ewgenius) [reply crates/secrets/src/aws_secrets_manager.rs]: fixed

File: crates/runtime/src/lib.rs [hunk 3]
```diff
@@ -93,6 +93,12 @@ pub enum Error {
     #[snafu(display("Unable to load secrets for data connector: {data_connector}"))]
     UnableToLoadDataConnectorSecrets { data_connector: String },
 
+    #[snafu(display("Unable to get secret for data connector: {source}"))]
```
Comment (phillipleblanc) [inline crates/runtime/src/lib.rs]:
```suggestion
    #[snafu(display("Unable to get secret for data connector {data_connector}: {source}"))]
```

File: crates/secrets/src/aws_secrets_manager.rs [hunk 4]
```diff
@@ -36,10 +36,17 @@ pub enum Error {
     UnableToVerifyAwsIdentity {
         source: SdkError<GetCallerIdentityError>,
     },
-    #[snafu(display("Unable to parse as JSON: {}", source))]
+
+    #[snafu(display("Unable to parse AWS secret as JSON: {}", source))]
```
Comment (phillipleblanc) [inline crates/secrets/src/aws_secrets_manager.rs]:
```suggestion
    #[snafu(display("Unable to parse AWS secret as JSON: {source}"))]
```

File: crates/secrets/src/aws_secrets_manager.rs [hunk 5]
```diff
@@ -36,10 +36,17 @@ pub enum Error {
...
     UnableToParseJson { source: serde_json::Error },
-    #[snafu(display("Invalid JSON format: JSON object is expected"))]
+
+    #[snafu(display("Invalid AWS secret value: JSON object is expected"))]
     InvalidJsonFormat {},
+
+    #[snafu(display("Unable to get AWS secret: {}", source))]
```
Comment (phillipleblanc) [inline crates/secrets/src/aws_secrets_manager.rs]:
```suggestion
    #[snafu(display("Unable to get AWS secret: {source}"))]
```

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**